modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,357 @@
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat("a", 3)
+----
+
+This returns `&quot;aaa&quot;`, but:
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(null, 3)
+----
+
+Throws an error because the value `null` cannot be assigned to the parameter that expects a value of type String.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(Null, 3)
+----
+
+In this case we are assigning to the first parameter a value that is the type Null (the type of this value
+is `Type&lt;Null&gt;`), and returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerces the value to be of type String (`&quot;Null&quot;`).
+
+Some other functions may have an overloaded signature for the function (the function has multiples definitions for different parameters types) that let it be called with `null`.
+
+For example, the `isNumeric` function from dw::core::Strings has the signature `isNumeric(String): Boolean` (receives a string and returns if that string is numeric) but it is also defined as `isNumeric(Null): Boolean`. +
+
+So if you call the function `isNumeric` with a null (`isNumeric(null)`) this does not throw an exception. At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used. The function returns false because of the definition of `isNumeric` for null as the argument.
+
+== Composite Types
+
+Composite types contains other values inside it. These composite types are Array, Object and Function.
+
+=== Array Type
+
+An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. For example, the syntax to define a variable with the type Array of `Number` is:
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.
+
+=== Object Type
+
+In order to define an object type there are two ways, either `Object` or `{}`, they both define an open object and don't specify any constraint to their key value pairs. Open and close objects are later defined in this section.
+
+The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
```

Comment:
I think you might need to use HTML code for the asterisk (. `&#42;`). It might not show up in the built HTML otherwise.